features.ts

```diff
@@ -0,0 +1,390 @@
+      }
+      const sceneJson = await sceneToJson(scene, state, req.session);
+
+      res.json({
+        error: false,
+        scene: sceneJson,
+      });
+
+    });
+
+  state.app.post(
+    "/features/queue/pop",
+    async (req: JwtRequest, res: Response) => {
+      const result = await tryPopFromFeatureQueue(state);
+
+      const queueDoc = result.value;
+      if (queueDoc === null) {
+        throw new Error(`Database failure, queue is missing`);
+      }
+
+      const id = queueDoc.scene_ids[0];
+      const scene = await state.scenes.findOne({ "_id": new ObjectId(id) });
+      if (scene === null) {
+        throw new Error(`Database consistency failure, feature queue missing scene ${id}`);
+      }
+      const sceneJson = await sceneToJson(scene, state, req.session);
+
+      res.json({
+        error: false,
+        scene: sceneJson
+      });
+    });
+
+  state.app.post(
+    "/features/queue/:id",
+    async (req: JwtRequest, res: Response) => {
+      const id = req.params.id;
+      const objectId = new ObjectId(id);
+      const scene = await state.scenes.findOne({ "_id": objectId });
+
+      if (scene === null) {
+        res.status(400).json({
+          error: true,
+          message: `Invalid scene id ${id}`
+        });
+        return;
+      }
+
+      const queueDoc = await state.featureQueue.findOne({ queue: true });
+      if (queueDoc === null) {
+        throw new Error(`Database failure, queue is missing`);
+      }
+
+      const inQueue = queueDoc.scene_ids.some(oid => oid.equals(objectId));
+      if (inQueue) {
+        res.json({
+          error: false,
+          message: `Scene ${id} is already in the queue`
+        });
+        return;
+      }
+
+      const updateResult = await state.featureQueue.updateOne(
+        { queue: true },
+        { "$push": { scene_ids: objectId } }
+      );
+
+      if (updateResult.modifiedCount === 0) {
+        res.status(500).json({
+          error: true,
+          message: `Error adding ${id} to queue`,
+        });
+      } else {
+        res.json({
+          error: false,
+          message: `Scene ${id} added to queue`,
+        });
+      }
+
+    });
+
+  state.app.delete(
+    "/features/queue/:id",
+    async (req: JwtRequest, res: Response) => {
+      const id = req.params.id;
+      const objectId = new ObjectId(id);
+      const queueDoc = await state.featureQueue.findOne({ queue: true });
+      if (queueDoc === null) {
+        throw new Error(`Database failure, queue is missing`);
+      }
+
+      const inQueue = queueDoc.scene_ids.some(oid => oid.equals(objectId));
+
+      if (!inQueue) {
+        res.json({
+          error: false,
+          message: `Scene ${id} was not in queue`,
+        });
+        return;
+      }
+
+      const newQueue = queueDoc.scene_ids;
+      newQueue.push(objectId);
+      const updateResult = await state.featureQueue.updateOne(
+        { queue: true },
+        { scene_ids: newQueue }
```

Comment:
Is there an operator that can delete an item from a list? That would potentially have more robust semantics in the off chance that someone is trying to update the queue from elsewhere simultaneously. But, again, we really don't expect that kind of contention for this API, and it's low-stakes if anything goes wrong, so it's not a big deal if this is the best available way to implement the deletion.

Reply:
Yep, we can use `$pull` - I used `$push` for adding an item, but completely forgot about `$pull` for some reason.